app/boards/arm/waveshare_rp2040_zero/waveshare_rp2040_zero.dts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2023 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/dts-v1/;
+
+#include <rpi_pico/rp2040.dtsi>
+#include "waveshare_rp2040_zero-pinctrl.dtsi"
+#include <freq.h>
+
+/ {
+	chosen {
+		zephyr,sram = &sram0;
+		zephyr,flash = &flash0;
+		zephyr,console = &uart0;
+		zephyr,code-partition = &code_partition;
+	};
+
+	xtal_clk: xtal-clk {
+		compatible = "fixed-clock";
+		clock-frequency = <12000000>;
+		#clock-cells = <0>;
+	};
+
+};
+
+&flash0 {
+	/* 2MB of flash minus the 0x100 used for
+	 * the second stage bootloader
+	 */
+	reg = <0x10000000 DT_SIZE_M(2)>;
+
+	partitions {
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		/*
+		 * Start at the beginning of usable flash, 8MB minus the
+		 * second stage space and the 16 KiB reserved for settings
+		 */
+		code_partition: partition@100 {
+			label = "code";
+			reg = <0x100 (DT_SIZE_M(2) - DT_SIZE_K(16))>;
+			read-only;
+		};
+
+		/*
+		 * The final 16 KiB is reserved for the application.
+		 * Storage partition may be used by FCB or LittleFS.
+		 */
+		storage_partition: partition@7fbe00 {
+			label = "storage";
+			reg = <0x007fbe00 DT_SIZE_K(16)>;
+		};
+	};
+};
+
+&uart1 {
+	current-speed = <115200>;
+	status = "okay";
```

Comment:
For ZMK boards, don't enable the UART, SPI, etc by default. This is a bit different from the upstream Zephyr board approach.